Prompt for signature on refresh [Fixes #220]

wackerow opened the pull request:
### Changes
- Adds new `localStorage.ts` utility functions to manage localStorage, and enable storage of chosen wallet type.
- Does not store any sensitive data in this unencrypted form
- Within `$web3` plugin, now stores the chosen wallet in localStorage on login and removes the chosen wallet on disconnect
- This also checks if there is an existing choice on page load, and if so, triggers a signature/login
- Adds `user` key to `$web3` export for storage, instead of returning the User object from method
- Moves the `loginUser` functionality to `App.vue` from `WalletModal.vue`, and decorates the function with `@Watch('$web3.user')` which will run this function whenever the stores user changes, returning early if no user found.
- While working in `App.vue`: Added functionality to clear the intervals that are set in the `created()` method when the component unmounts to prevent memory leaks. 
- Updates wallet login message

### tl;dr:
User will now be prompted for a signature on page refresh if they've previously logged in using a chosen wallet provider. Upon "disconnecting" this choice is removed from localStorage, and in this case refreshing the page will not prompt user.

wackerow committed:
adds `connection` utility

Adds localStorage getter/setter/remover

wackerow committed:
checks localStorage for existing wallet provider

Adds `user` key to `$web3` export for storage, instead of returning the User object from method

wackerow committed:
refactor loginUser logic into App.vue

Adds @Watch decorator to `loginUser()` which is triggered whenever `$web3.user` changes.

wackerow committed:
clear App.vue intervals on unmount

wackerow committed:
removes console log

wackerow committed:
Merge branch 'develop' into feat/sign-on-refresh

wackerow committed:
remove stored wallet type on failed connection

If a user cancels the signature, or another error occurs when attempting to connect, the stored wallet type will be cleared. This prevents the signature from repeatedly and automatically being prompted on refresh after the user may have canceled intentionally.

wackerow committed:
fixed handling for wallet cancellation

also extracted string into constant

pettinarip commented on vue-app/src/utils/connection.ts:
```diff
@@ -0,0 +1,26 @@
+
+export function lsRemove(key: string) {
+  return localStorage.removeItem(lsGetKey(key))
+}
```
Everthing in here is related to local storage manipulation. I would call this file `localStorage.ts`.

pettinarip commented on vue-app/src/utils/connection.ts:
```diff
@@ -0,0 +1,26 @@
+import pkg from '@/../package.json'
+
+function lsGetKey(key: string) {
+  return `${pkg.name}.${key}`
```
cool!

pettinarip commented on vue-app/src/plugins/Web3/index.ts:
```diff
@@ -111,6 +120,15 @@ export default {
+    // If previous provider was found, initiate connection.
+    if (alreadyConnectedProvider) {
+      const response = plugin.connectWallet(alreadyConnectedProvider)
+      if (response.state !== 'fulfilled') {
```
Wondering what `state` does in this case since we are not `await`ing the promise. I mean, in which other cases are we going to face a different value than `'fulfilled'`?

wackerow replied:
@pettinarip So this only returns `'fulfilled'` is the user actually signs... Adding this basically is checking if the user cancels the prompt. If they do, it will then clear the localStorage.

ie: If the user clicks "cancel" on MetaMask -> returns `response.state = 'pending'`... 

When I tried to wrap this in `await` syntax, the app does not respond and update with the logged in user. Clicking "Connect" and signing just leads to the same screen with the "Connect" button still showing. Keeping it synchronous as above seems to work properly.

pettinarip replied:
I just tested this and understood what you want to do.

Actually, checking the promise `state` prop is not going to tell you if the user has accepted or not the prompt. That state is part of the Promise lifecycle (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise). When the Promise is `resolve`d or `reject`ed, the state variable is going to change.

So, in this case, `response.state` is always going to be `pending` because we are not waiting for it to end. Basically, the condition is always going to be true.

Regarding what you want to achieve here I think we could do the following:
```ts
if (alreadyConnectedProvider) {
  // What we really want to do is to clean the `alreadyConnectedProvider` provider, not do a complete disconnection.
  // This is mainly to avoid future problems if we add more logic in the `disconnectWallet` method
  lsRemove(CONNECTED_PROVIDER) 
  plugin.connectWallet(alreadyConnectedProvider)
}
```

samajammin replied:
Agreed - thanks for the explanation @pettinarip!

pettinarip commented:
Looks great @wackerow. I left a couple of small comments.

I'm wondering if we should consider the following case:
1- User connects with MM
2- Refreshes the page
3- Rejects the signature
4- Tries to connect with a different wallet, WalletConnect
5- Should we manually disconnect from the previous wallet (MM)? or doesn't it hurt us to leave it as connected?

wackerow commented:
@pettinarip See the other comment about this, as rejecting the signature should basically start someone back at square one, with no wallet preference saved. Clicking connect at this point will offer the MetaMask or WalletConnect option again.

wackerow committed:
rename connection util to localStorage

pettinarip commented:
> @pettinarip See the other comment about this, as rejecting the signature should basically start someone back at square one, with no wallet preference saved. Clicking connect at this point will offer the MetaMask or WalletConnect option again.

Got it. Yes, looks great that approach. To remove the `alreadyConnectedProvider` from the ls before forcing the connection again.

wackerow committed:
updates wallet login message

samajammin commented on vue-app/src/App.vue:
```diff
@@ -96,6 +103,28 @@ export default class App extends Vue {
     await this.$store.dispatch(LOAD_RECIPIENT_REGISTRY_INFO)
   }
 
+  beforeDestroy() {
```
👍

samajammin commented on vue-app/src/App.vue:
```diff
@@ -96,6 +103,28 @@ export default class App extends Vue {
+    }
+  }
+
+  @Watch('$web3.user')
```
So is this how class based components adds watchers?

wackerow replied:
https://github.com/kaorun343/vue-property-decorator#Watch 👍🏻

samajammin commented on vue-app/src/App.vue:
```diff
@@ -96,6 +103,28 @@ export default class App extends Vue {
+  loginUser = async () => {
+    if (!this.$web3.user) return
+
+    this.$store.commit(SET_CURRENT_USER, this.$web3.user)
```
Won't this `SET_CURRENT_USER` overwrite all other user data that $web3 doesn't have (such as `isRegistered` & `brightId`)?

samajammin replied:
I suppose that's ok since we subsequently dispatch `LOAD_USER_INFO`? Is that the thinking here?

pettinarip replied:
Well, you are right but we shouldn't reach the case where we overwrite something by mistake. Mainly because this function is only going to be called when `$web3.user` changes. That happens in two moments:

1. When `disconnectWallet` is called: the top guard clause is for this case.
2. When `connectWallet` is called: we are going to overwrite the user but we are calling the rest of the data that completes the user state.

samajammin commented on vue-app/src/utils/localStorage.ts:
```diff
@@ -0,0 +1,26 @@
+import pkg from '@/../package.json'
+
+function lsGetKey(key: string) {
+  return `${pkg.name}.${key}`
```
Could you explain why we use `pkg.name` - just to give our keys a namespace?

wackerow replied:
Yes, this was just to ensure a unique namespace

samajammin reviewed (changes requested):
This looks amazing! 🤩 

Agree w/ @pettinarip's commet re: the `alreadyConnectedProvider` check - let's make that change before merging.

samajammin commented:
Tried it on both MetaMask & WalletConnect - it's looking super slick!

wackerow committed:
update rejection logic

If there was an existing provider, first clear it from ls, then prompt connection with the previously chosen provider. If accepted, this will re-save in ls, otherwise it will already be erased.
